[nomerge] optimise the addition of immutable HashMap

mkeskells opened the pull request:
Note - for 2.12 only
We can duplicate the functionality to 2.13, but I think it is with a seperate PR as 2.13 collections are differently structured

Prefer to use `merge` when we can
optimised the builder to allow `++` directly

small collection of other optimisations

need some additional tests for the edge cases introduced

retronym commented:
Minimization of the bootstrap failure:

```
scala> val breaker = collection.breakOut[scala.collection.immutable.Set[Int], Int, scala.collection.immutable.BitSet](scala.collection.immutable.BitSet.canBuildFrom)
breaker: scala.collection.generic.CanBuildFrom[scala.collection.immutable.Set[Int],Int,scala.collection.immutable.BitSet] = scala.collection.generic.BitSetFactory$$anon$1@3dcb28de

scala> val f = java.lang.Math.abs _
f: Int => Int = $$Lambda$1331/1191964420@1ab08e83

scala> Set[Int]().map[Int, scala.collection.immutable.BitSet](f)(breaker)
java.lang.ClassCastException: scala.collection.immutable.Set$EmptySet$ cannot be cast to scala.collection.BitSet
  at scala.collection.generic.BitSetFactory$$anon$1.apply(BitSetFactory.scala:36)
  at scala.collection.TraversableLike.builder$1(TraversableLike.scala:233)
  at scala.collection.TraversableLike.map(TraversableLike.scala:237)
  at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
  at scala.collection.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:51)
  at scala.collection.SetLike.map(SetLike.scala:104)
  at scala.collection.SetLike.map$(SetLike.scala:104)
  at scala.collection.AbstractSet.map(Set.scala:51)
```

mkeskells commented:
rebased to resolve expecte conflict with https://github.com/scala/scala/pull/8467

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -137,6 +137,79 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
+    cbf match {
+      case w: WrappedCanBuildFrom[_,_,_] =>
+        val unwrapped = w.unwrap
+        (unwrapped eq HashMap.canBuildFrom) || (unwrapped eq Map.canBuildFrom)
```
I'm not sure if you can ever get multiple levels of wrapping but `isCompatibleCBF(w.unwrap)` would also be more concise and avoid repeating the test

mkeskells replied:
done

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -137,6 +137,79 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
+
+  override def ++[C >: (A, B), That](that: GenTraversableOnce[C])(implicit bf: CanBuildFrom[HashMap[A, B], C, That]): That =
+    addImpl(that, bf)
+  private def addImpl[C >: (A, B), That](that: GenTraversableOnce[C], bf: CanBuildFrom[HashMap[A, B], C, That]): That = {
```
Is the separate method necessary? Couldn't you put the implementation directly into the second `++` method?

mkeskells replied:
if I chnage  to 
```
  override def ++[B1 >: B](xs: GenTraversableOnce[(A, B1)]): Map[A, B1] = ++(xs)(HashMap.canBuildFrom[A, B1])
```
I get a compile error
```
Error:(166, 102) type mismatch;
 found   : scala.collection.generic.CanBuildFrom[scala.collection.immutable.HashMap.Coll,(A, B1),scala.collection.immutable.HashMap[A,B1]]
    (which expands to)  scala.collection.generic.CanBuildFrom[scala.collection.immutable.HashMap[_, _],(A, B1),scala.collection.immutable.HashMap[A,B1]]
 required: A
  override def ++[B1 >: B](xs: GenTraversableOnce[(A, B1)]): Map[A, B1] = ++(xs)(HashMap.canBuildFrom[A, B1])
```

szeiger replied:
It works if you specify the type parameters:
```
  override def ++[B1 >: B](xs: GenTraversableOnce[(A, B1)]): Map[A, B1] = ++[(A, B1), Map[A, B1]](xs)(HashMap.canBuildFrom[A, B1])

  override def ++[C >: (A, B), That](that: GenTraversableOnce[C])(implicit bf: CanBuildFrom[HashMap[A, B], C, That]): That = {
```
The difference must be due to `++` being overloaded. It's strange that it fails with this type error rather than an ambiguous overload error.

mkeskells replied:
done - thanks for the type magic :-)

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -137,6 +137,79 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
+      else if (that.isEmpty) this.asInstanceOf[That]
+      else that match {
+        case thatHash: HashMap[A, B] =>
+          //default Merge prefers to keep than replace
+          //so we merge from thatHash
+          (thatHash.merged(this) (null) ).asInstanceOf[That]
```
Is the handling of key and value identities here and in `++:` below compatible with the current default (and new fallback) implementation consisting of successive `+` calls?

mkeskells replied:
I believe so - that was my intention

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -137,6 +137,79 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
+          var result: HashMap[A, B] = this.asInstanceOf[HashMap[A, B]]
+          that foreach { case kv: (A, B) =>
+            val key = kv._1
+            result = result.updated0(key, computeHash(key), 0, kv._2, kv, HashMap.liftMerger[A,B](null))
```
Is the default merger (aka `liftMerger(null)`) different from a `null` merger (as used when calling `+`)?

mkeskells replied:
yes - `liftMerger(null)` uses a merger with `retainIdentical = true` 

```
  private def liftMerger[A1, B1](mergef: MergeFunction[A1, B1]): Merger[A1, B1] =
    if (mergef == null) defaultMerger.asInstanceOf[Merger[A1, B1]] else liftMerger0(mergef)

  private val defaultMerger : Merger[Any, Any] = new Merger[Any, Any] {
    override def apply(a: (Any, Any), b: (Any, Any)): (Any, Any) = a
    override def retainIdentical: Boolean = true
    override val invert: Merger[Any, Any] = new Merger[Any, Any] {
      override def apply(a: (Any, Any), b: (Any, Any)): (Any, Any) = b
      override def retainIdentical: Boolean = true
      override def invert = defaultMerger
    }
  }
```

this means that the merge code can ignore structurally shared trees

other than that the behaviour is , or should be, identical

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -137,6 +137,79 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
+            result = result.updated0(key, computeHash(key), 0, kv._2, kv, HashMap.liftMerger[A,B](null))
+          }
+          result.asInstanceOf[That]
+      }
```
This whole branch appears to be identical to the one from the other `++:` method above. It could be factored out and shared.

mkeskells replied:
done - and removed the ObjectRef
TODO - should cope with the Maps that now have `forEachEntry`

mkeskells replied:
did the forEachEntry change, and copied most of that to another PR to reduce the noise in the one

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -169,8 +251,18 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
 
+  override def newBuilder[A, B]: mutable.Builder[(A, B), HashMap[A, B]] = new HashMapBuilder[A,B]
+  private class HashMapBuilder[A, B] extends mutable.MapBuilder[A, B, HashMap[A, B]](HashMap.empty) {
+    //not sure if this should be part of MapBuilder
```
Probably, but the override wouldn't be binary compatible and there's a risk of breaking existing implementations that make `++` in `Map` call `MapBuilder`.

mkeskells replied:
Maybe we could do something in 2.15, when that branch starts

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -517,65 +651,120 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
+              // condition below is due to 2 things:
+              // 1) no unsigned int compare on JVM
+              // 2) 0 (no lsb) should always be greater in comparison
+              if (unsignedCompare(thislsb - 1, thatlsb - 1)) {
```
Off-topic: I wonder if `unsignedCompare` (for `Int`) was ever benchmarked against a more straight-forward version that performs unsigned conversions to `Long`. It's also confusingly named (compare? for what?)

mkeskells replied:
I think (based in the HashSet changes, this can probably go from this implementation later

mkeskells replied:
removed reference to unsignedCompare from HashMap. There are still some from HashSet until https://github.com/scala/scala/pull/8589 merges

szeiger commented on src/library/scala/collection/immutable/Map.scala:
```diff
@@ -108,6 +108,28 @@ object Map extends ImmutableMapFactory[Map] {
     def + [V1](kv: (Any, V1)): Map[Any, V1] = updated(kv._1, kv._2)
+    override def ++[V1 >: Nothing](xs: GenTraversableOnce[(Any, V1)]): Map[Any, V1] = addImpl(xs)
+    override def ++[B >: (Any, Nothing), That](that: GenTraversableOnce[B])(implicit bf: CanBuildFrom[Map[Any, Nothing], B, That]): That = super.++(that)
+    private def addImpl[B >: (Any, Nothing), That](that: GenTraversableOnce[B])(implicit bf: CanBuildFrom[Map[Any, Nothing], B, That]): That = {
```
It looks like this is only called from the first `++` above where the cbf instance is statically known.

mkeskells replied:
doh - changed to be similar to HashMap `++` and `addImpl`

szeiger commented on src/library/scala/collection/package.scala:
```diff
@@ -94,10 +96,16 @@ package object collection {
+    def apply(from: From) = wrapped.apply()
+
+    def apply() = wrapped.apply()
+    final def unwrap:  CanBuildFrom[_, T, To] = wrapped
```
Why do we need this method that drops the `From` type instead of making `wrapped` a `val` and keeping `Nothing`?

mkeskells replied:
no idea. Changed it to be a val

mkeskells commented:
addressed some of @szeiger comments
rebased - (removed the mima changes - easier to start again on those from the build)

mkeskells committed:
add a benchmark

mkeskells committed:
Improve performance of HashMap merge and ++ (mostly).

use retainIdentical where appropriate to combine trees
Learning from the optimisations of HashSet bulk operations, port some of that where appropriate to HashMap
 - mostly around HashTrie merge
 - remove more allocations

use System.arraycopy rather then Array.copy for performance
add fast path for some trivial cases, e.g. empty
add fast path to EmptyMap ++ Map, as builders use this path
add allocation testing
add test to ensure simple cases preserve identity on merge

mkeskells commented:
rebased and squashed

mkeskells committed:
add HasForeachEntry

mkeskells committed:
remove a couple of forwarding method

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -209,7 +210,17 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
+          val merger = HashMap.liftMerger[A, B](null)
+
+          override def apply(key: A, value: B): Unit = {
+            result = result.updated0(key, computeHash(key), 0, value, null, merger)
```
Is this really better than using `foreach`? Now you don't get a tuple anymore (even when the original map has one) so you have to compute a new one later on demand.

mkeskells replied:
Its a trade off. maybe the foreachEntry should have a tuple if it has one?
It depends if you generate the tuple, and how many Maps actually have the tuple I think

mkeskells replied:
the only classes that implement `HasForEachEntry` are 
`HashMap` - covered previously in matches
`ListMap` - doesn't  maintain tuples
`SortedMap`/`RedBlackTree` - doesn't maintain tuples 

which collections types are you concerned with, when we have the existing tuple and lose it?

szeiger commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -322,27 +333,31 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
+            val nkv = merger(current, if (kv != null) kv else (key, value))
+            if ((current eq nkv) || (
+              (key.asInstanceOf[AnyRef] eq nkv._1.asInstanceOf[AnyRef]) &&
+                (value.asInstanceOf[AnyRef] eq nkv._2.asInstanceOf[AnyRef]))) this
```
Doesn't this need to be guarded by `merger.retainIdentical` but can be removed because it was already covered by the previous branch anyway?

mkeskells replied:
I don't see why. 
If the merge generates a tuple which is identical in all respects then we don't need to create a new `HashMap1`, when this is a clone and identical in all respects to `this`
What user visible difference is there? 
Non-functionally there is a change because not having this would cause the creation of a new `HashMap1` and all of the parent structures, and the loss of structural sharing

mkeskells replied:
I think I see the confusion - there was a bug here
key and values are updated to be this.key and this.value - no referring to the parameters

szeiger replied:
Ah, that explains it. I was looking for the context of the code again after your previous comment but github didn't have it anymore so I looked at the latest version and everything there looked fine.

retronym commented:
Pushed a commit to rebase atop the latest 2.12.x (which has `HasForEach`) and to squash the last two commits.

retronym commented:
/synch

retronym commented:
/nothingtoseehere

(Clearing spurious "cla Expected — Waiting for status to be reported ")

szeiger commented:
LGTM but it's hard to judge the correctness of all the new code so I'd prefer to get a 2nd approval from @retronym before merging.

SethTisue commented:
after this PR, scalariform has test failures in the community build: https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/5532/artifact/logs/scalariform-build.log

regression?

retronym commented:
Fixed in https://github.com/scala/scala/pull/8764
